src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala

```diff
@@ -450,6 +471,34 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
     }
   }
 
+  private def isConstrainable(resource: Resource, actions: Set[ResourceAction]): Boolean = {
+    resourceTypes.get(resource.resourceTypeName) match {
+      case Some(resourceType) => {
+        for {
+          actionPattern <- resourceType.actionPatterns
+          action <- actions
+        } yield actionPattern.authDomainConstrainable && actionPattern.matches(action)
+      }.contains(true)
+      case None => false // we really shouldn't get here as this means the resource is invalid, but leaving it in for completeness
+    }
+  }
+
+  private def calculateIntersectionGroup(resource: Resource, policy: AccessPolicy): Future[Set[WorkbenchUserId]] = {
+    for {
+      groups <- accessPolicyDAO.loadResourceAuthDomain(resource)
+      authDomain <- Future.traverse(groups)(groupName =>
+        directoryDAO.listFlattenedGroupUsers(groupName))
+      policyMembers <- Future.traverse(policy.members)(groupName => groupName match {
+        case g: WorkbenchGroupName => directoryDAO.listFlattenedGroupUsers(g)
+        case u: WorkbenchUserId => Future.successful(Set(u))
+        case p: PetServiceAccountId => Future(Set(p.userId))
+        case _ => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Bad policy found"))
+      })
+    } yield {
+      (authDomain + policyMembers.flatten).reduce(_.intersect(_))
```

Comment:
if `authDomain + policyMembers.flatten` is empty, `reduce` will throw exception. Will that be an issue? When it's empty, shall we just return empty set or we really want to propagate the error?